File: core/state/stateclient/StateClient.go
package stateclient

import (
	"github.com/hiveot/hub/core/state/stateapi"
	"github.com/hiveot/hub/lib/hubclient"
	"github.com/hiveot/hub/lib/ser"
)

// StateClient is a marshaller for service messages using a provided hub connection.
// This uses the default serializer to marshal and unmarshal messages.
type StateClient struct {
	// ID of the service that handles the requests
	agentID string
	// State storage capability
	capID string
	// Connection to the hub
	hc hubclient.IHubClient
}

// Delete removes the record with the given key.
func (cl *StateClient) Delete(key string) error {

	req := stateapi.DeleteArgs{Key: key}
	_, err := cl.hc.PubRPCRequest(
		cl.agentID, cl.capID, stateapi.DeleteMethod, &req, nil)
	return err
}

// Get reads and unmarshals the record with the given key.
// If the key doesn't exist this returns an empty record.
func (cl *StateClient) Get(key string, record interface{}) (found bool, err error) {

	req := stateapi.GetArgs{Key: key}
	resp := stateapi.GetResp{}
	_, err = cl.hc.PubRPCRequest(
		cl.agentID, cl.capID, stateapi.GetMethod, &req, &resp)
	if err != nil {
		return false, err
	}
	if resp.Found {
		err = ser.Unmarshal(resp.Value, record)
	}
	return resp.Found, err
}

// GetMultiple reads multiple records with the given keys.
// This marshalling and unmarshalling is up to the caller.
func (cl *StateClient) GetMultiple(keys []string) (values map[string][]byte, err error) {

	req := stateapi.GetMultipleArgs{Keys: keys}
	resp := stateapi.GetMultipleResp{}
	_, err = cl.hc.PubRPCRequest(
		cl.agentID, cl.capID, stateapi.GetMultipleMethod, &req, &resp)
	if err != nil {
		return nil, err
	}
	return resp.KV, err
}

// Set marshals and writes a record
func (cl *StateClient) Set(key string, record interface{}) error {
	value, err := ser.Marshal(record)
	if err != nil {
		return err
	}
	req := stateapi.SetArgs{Key: key, Value: value}
	_, err = cl.hc.PubRPCRequest(
		cl.agentID, cl.capID, stateapi.SetMethod, &req, nil)
	return err
}

// SetMultiple writes multiple record
func (cl *StateClient) SetMultiple(kv map[string][]byte) error {
	req := stateapi.SetMultipleArgs{KV: kv}
	_, err := cl.hc.PubRPCRequest(
		cl.agentID, cl.capID, stateapi.SetMultipleMethod, &req, nil)
	return err
}

// NewStateClient returns a client to access state
//
//	hc is the hub client connection to use.
func NewStateClient(hc hubclient.IHubClient) *StateClient {
	agentID := stateapi.ServiceName
	cl := StateClient{
		hc:      hc,
		agentID: agentID,
		capID:   stateapi.StorageCap,
	}
	return &cl
}

File: libjs/src/clients/stateclient.ts
import {IAgentClient} from "@hivelib/hubclient/IAgentClient";
import {ActionProgress} from "@hivelib/hubclient/ActionProgress";

// duplicated from stateapi.go
const AgentID = "state"      // the state binding agentID
const StateStoreID = "store" // thingID of the storage service
const DeleteMethod = "delete"
const GetMethod = "get"
const GetMultipleMethod = "getMultiple"
const SetMethod = "set"
const SetMultipleMethod = "setMultiple"


// Marshaller for invoke state service method using the given hub client
export class StateClient {
    hc: IAgentClient
    serviceID: string // digitwin ID of the state service

    constructor(hc: IAgentClient) {
        this.hc = hc
        // this.serviceID = makeDigiTwinThingID(AgentID, StateStoreID)
        this.serviceID = "dtw:"+AgentID+":"+ StateStoreID
    }

    // Delete a key
    async Delete(key: string) {
        let args = {
            key: key
        }
        return await this.hc.rpc(this.serviceID, DeleteMethod, args)
    }

    // Get the value of a key
    async Get(key: string): Promise<{ value: string, found: boolean }> {
        let args = {
            key: key
        }
        type RespType = {
            key: string
            found: boolean
            value: string
        }
        let resp:RespType
        try {
            resp = await this.hc.rpc(this.serviceID, GetMethod, args)
        } catch(e) {
            console.error("state.Get error",e)
            throw("state.Get error"+e)
        }
        return { value: resp.value, found: resp.found }
    }

    // Get a map of key:value 
    async GetMultiple(keys: string[]): Promise<{ [index: string]: string }> {
        let args = {
            keys: keys
        }
        type RespType = {
            kv: { [index: string]: string }
        }
        let resp: RespType = await this.hc.rpc(this.serviceID, GetMultipleMethod, args)

        return resp.kv
    }

    // Set the value of a key
    async Set(key: string, data: string):Promise<ActionProgress> {
        let args = {
            key: key,
            value: data
        }
        let stat = await this.hc.rpc(this.serviceID, SetMethod, args)
        return stat
    }

    // Set multiple values at once
    async SetMultiple(kv: { [index: string]: string }) {
        let args = {
            kv: kv
        }
        return await this.hc.rpc(this.serviceID, SetMultipleMethod, args)
    }
}

